File: Storage/LinkedList.py
from Link import Link

class LinkedList:
    def __init__(self, elem=None):
        # Top has to be Node object
        self.top = Link(elem)
        self.size = 1

    def add_Elem(self, elem):
        temp = self.top
        # Loop until the last node
        while temp.getNext() is not None:
            temp = temp.getNext()
        temp.setNext(Link(elem))
        self.size += 1

    def add_Elem_pos(self, pos, elem):
        # Stop if pos is greater than the linkedlist size
        if pos > self.size: return None
        new_node = Link(elem)
        # If Add node at top
        if pos == 0:
            new_node.setNext(self.top)
            self.top = new_node
        temp = self.top
        it = 0
        # Loop until the previous node of pos
        while temp.getNext() is not None and it < pos:
            if it+1 == pos: # If the next node is pos
                last_node = temp
                next_node = temp.getNext()
                new_node.setNext(next_node)
                last_node.setNext(new_node)
                self.size += 1
                return
            temp = temp.getNext()
            it += 1
        # Add node at tail
        if it+1 == pos:
            temp.setNext(new_node)
            self.size += 1

    def remove(self, pos):
        # Stop if pos is greater than the linkedlist size
        if pos > self.size: return
        # Remove the top
        if pos == 0:
            if self.top.getNext() is not None:
                self.top = self.top.getNext()
            else:
                self.top = None
            self.size -= 1
            # Once is done, return
            return
        temp = self.top
        it = 0
        # Loop until the previous node of pos
        while temp.getNext() is not None and it < pos:
            if it+1 == pos:
                # If this is the tail
                if temp.getNext() is None:
                    temp.setNext(None)
                else:
                    # If this is not tail; If the next node is node is fine
                    temp.setNext(temp.getNext().getNext())
                # Once the operation is done, -1 on size
                self.size -= 1
                return
            temp = temp.getNext()
            it += 1

    def get(self, pos):
        # If linkedlist is empty
        if self.size == 0: return None
        if pos > self.size: return None
        temp = self.top
        it = 0
        while temp.getNext() is not None and it < pos:
            temp = temp.getNext()
            it += 1
        return temp.elem

    def getSize(self) -> int:
        return self.size

    def printList(self):
        # Helper function to print the whole linkedlist
        temp = self.top
        if temp is None:
        	print('This is Empty')
        	return
        print(temp.getElem())
        while temp.getNext() is not None:
            temp = temp.getNext()
            print(temp.getElem())

File: Storage/Ruby/LinkedList.rb
require_relative 'Link'


class LinkedList
	def initialize(elem=nil)
		@top = Link.new(elem)
		@size = 1
	end

	def add_Elem(elem)
		temp = @top
		# Loop until the last node
		while !temp.next_Link.nil? do
			temp = temp.next_Link
		end # end while
		temp.next_Link = Link.new(elem)
		@size += 1
	end # end func

	def add_Elem_pos(pos, elem)
		# Stop if pos is greater than the linkedlist size
		if pos > @size then return end
		new_node = Link.new(elem)
		# If Add node at top
		if pos == 0
			then new_node.next_Link = @top
			@top = new_node
			return
		end # end if
		temp = @top
		it = 0
		# Loop until the previous node of pos
		while !temp.next_Link.nil? && it < pos do
			# If the next node is pos
			if it+1 == pos
				last_node = temp
				next_node = temp.next_Link
				new_node.next_Link = next_node
				last_node.next_Link = new_node
				@size += 1
				return
			end # end if
			temp = temp.next_Link
			it += 1
		end # end while
		# Add node at tail
		if it+1 == pos then
			temp.next_Link = new_node
			@size += 1
		end # end if
	end #end func

	def remove(pos)
		# Stop if pos is greater than the linkedlist size
		if pos > @size then return end
		# If removing the top
		if pos == 0
			unless @top.next_link.nil?
				@top.next_Link = nil
			else
				@top = @top.next_Link
			end # end unless
			@size -= 1
			return # return once it is done
		end # end if
		temp = @top
		it = 0
		# Loop until the previous node of pos
		while !temp.next_Link.nil? do
			if it+1 == pos
				then
					# If this is tail
					if temp.next_Link.nil? then temp.next_Link = nil 
					else
						# if this is not tail; if the next node is node is fine
						temp.next_Link = temp.next_Link.next_Link
					end # end if
					@size -= 1
					return
			end # end if
			temp = temp.next_Link
			it += 1
		end # end while
	end # end func

	def get(pos)
		# If linkedlist is empty
		if @size == 0 then return nil end
		# If pos is greater than size
		if pos > @size then return nil end
		temp = @top
		it = 0
		while !temp.next_Link.nil? and it < pos do
			temp = temp.next_Link
			it += 1
		end # end while
		return temp.elem
	end # end func

	def getSize
		return @size
	end # end func

	def printList
		# Helper function to print the whole linkedlist
		temp = @top
		if temp.nil? 
			then puts"This is Empty" 
				return 
		end # end if

		puts temp.elem
		while !temp.next_Link.nil? do
			temp = temp.next_Link
			puts temp.elem
		end # end while
	end # end func
end # end class